src/Messages/Attachments/OWSAttachmentsProcessor.m

```diff
@@ -233,22 +233,64 @@ - (void)downloadFromLocation:(NSString *)location
     manager.responseSerializer = [AFHTTPResponseSerializer serializer];
     manager.completionQueue    = dispatch_get_main_queue();
 
+    // We want to avoid large downloads from a compromised or buggy service.
+    const long kMaxDownloadSize = 150 * 1024 * 1024;
     // TODO stream this download rather than storing the entire blob.
-    [manager GET:location
-      parameters:nil
-        progress:nil // TODO show some progress!
-         success:^(NSURLSessionDataTask *_Nonnull task, id _Nullable responseObject) {
-             if (![responseObject isKindOfClass:[NSData class]]) {
-                 DDLogError(@"%@ Failed retrieval of attachment. Response had unexpected format.", self.tag);
-                 NSError *error = OWSErrorMakeUnableToProcessServerResponseError();
-                 return failureHandler(task, error);
-             }
-             successHandler((NSData *)responseObject);
-         }
-         failure:^(NSURLSessionDataTask *_Nullable task, NSError *_Nonnull error) {
-             DDLogError(@"Failed to retrieve attachment with error: %@", error.description);
-             return failureHandler(task, error);
-         }];
+    __block NSURLSessionDataTask *task = nil;
+    // We only need to check the content length header once.
+    __block BOOL hasCheckedContentLength = NO;
+    task = [manager GET:location
+        parameters:nil
+        progress:^(NSProgress *_Nonnull progress) {
+            if (!hasCheckedContentLength) {
+                NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)task.response;
+                if ([httpResponse isKindOfClass:[NSHTTPURLResponse class]]) {
+                    NSDictionary *headers = [httpResponse allHeaderFields];
+                    if ([headers isKindOfClass:[NSDictionary class]]) {
+                        NSString *contentLength = headers[@"Content-Length"];
+                        if ([contentLength isKindOfClass:[NSString class]]) {
+                            if (contentLength.longLongValue > 0) {
+                                if (contentLength.longLongValue <= kMaxDownloadSize) {
+                                    // This response has a valid content length that is less
+                                    // than our max download size.  Proceed with the download.
+                                    hasCheckedContentLength = YES;
+                                    return;
+                                }
+                            }
+                        }
+                    }
+                }
```

Comment:
nit: This is nested pretty deep, to the point that I guessed wrong twice when trying to eyeball which condition this bracket closes. I think the above could be made much clearer by using guard statements and some kind of early return (in this case probably utilizing a block that has an assertion failure and a `[task cancel]`).

Also it would allow you to log a specific error at the point of the failing test case. =)